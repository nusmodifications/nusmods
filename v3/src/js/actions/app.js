// @flow

import type { FSA } from 'types/redux';

export const SET_ONLINE_STATUS = 'SET_ONLINE_STATUS';
export function setOnlineStatus(isOnline: boolean): FSA {
  return {
    type: SET_ONLINE_STATUS,
    payload: { isOnline },
  };
}

export const TOGGLE_FEEDBACK_MODAL = 'TOGGLE_FEEDBACK_MODAL';
export function toggleFeedback(): FSA {
  return {
    type: TOGGLE_FEEDBACK_MODAL,
    payload: null,
  };
}
