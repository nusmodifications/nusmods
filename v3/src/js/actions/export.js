// @flow
import domtoimage from 'dom-to-image';
import { iCalForTimetable } from 'utils/ical';

export const DOWNLOAD_AS_JPEG = 'DOWNLOAD_AS_JPEG';
export function downloadAsJpeg(domElement: Element) {
  return (dispatch: Function) => {
    dispatch({
      type: `${DOWNLOAD_AS_JPEG}_PENDING`,
    });

    const style = { margin: '0', marginLeft: '-0.25em' };
    return domtoimage.toJpeg(domElement, { bgcolor: '#fff', style })
      .then((dataUrl) => {
        const link = document.createElement('a');
        link.download = 'timetable.jpeg';
        link.href = dataUrl;
        link.click();
        dispatch({
          type: `${DOWNLOAD_AS_JPEG}_SUCCESS`,
        });
      })
      .catch((err) => {
        console.error(err);
        dispatch({
          type: `${DOWNLOAD_AS_JPEG}_FAILURE`,
        });
      });
  };
}

export const DOWNLOAD_AS_ICAL = 'DOWNLOAD_AS_ICAL';
export function downloadAsIcal(semester, timetable, moduleData) {
  const results = iCalForTimetable(semester, timetable, moduleData);
  return {
    type: DOWNLOAD_AS_ICAL,
    payload: results,
  };
}
