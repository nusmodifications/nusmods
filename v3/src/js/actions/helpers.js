// @flow

import _ from 'lodash';

import type { FSA } from 'types/redux';

export const RESET_REQUEST_STATE: string = 'RESET_REQUEST_STATE';
export function resetRequestState(domain: string | Array<string>): FSA {
  return {
    type: RESET_REQUEST_STATE,
    payload: _.isArray(domain) ? domain : [domain],
  };
}

export const RESET_ERROR_STATE: string = 'RESET_ERROR_STATE';
export function resetErrorState(domain: string | Array<string>): FSA {
  return {
    type: RESET_ERROR_STATE,
    payload: _.isArray(domain) ? domain : [domain],
  };
}

export const RESET_ALL_STATE: string = 'RESET_ALL_STATE';
export function resetAllState(): FSA {
  return {
    type: RESET_ALL_STATE,
  };
}
