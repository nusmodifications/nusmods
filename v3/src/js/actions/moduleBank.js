// @flow
import type {
  ModuleCode,
} from 'types/modules';
import type { FSA } from 'types/redux';

import { API_REQUEST } from 'middlewares/requests-middleware';
import NUSModsApi from 'apis/nusmods';

export const FETCH_MODULE_LIST: string = 'FETCH_MODULE_LIST';
export function fetchModuleList(): FSA {
  return {
    type: FETCH_MODULE_LIST,
    payload: {
      method: 'GET',
      url: NUSModsApi.moduleListUrl(),
    },
    meta: {
      [API_REQUEST]: true,
    },
  };
}

export const FETCH_MODULE: string = 'FETCH_MODULE';
export function fetchModule(moduleCode: ModuleCode): FSA {
  return {
    type: FETCH_MODULE,
    payload: {
      method: 'GET',
      url: NUSModsApi.moduleDetailsUrl(moduleCode),
    },
    meta: {
      [API_REQUEST]: true,
    },
  };
}
