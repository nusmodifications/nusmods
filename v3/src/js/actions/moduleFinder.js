// @flow

import type { FSA } from 'types/redux';

export const SEARCH_MODULES = 'SEARCH_MODULES';
export function searchModules(searchTerm: string): FSA {
  return {
    type: SEARCH_MODULES,
    payload: { searchTerm },
  };
}

export const RESET_MODULE_FINDER = 'RESET_MODULE_FINDER';
export function resetModuleFinder(): FSA {
  return {
    type: RESET_MODULE_FINDER,
    payload: null,
  };
}
