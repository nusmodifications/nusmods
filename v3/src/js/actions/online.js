// @flow

import type { FSA } from 'types/redux';

export const SET_ONLINE_STATUS = 'SET_ONLINE_STATUS';
export function setOnlineStatus(isOnline: boolean): FSA {
  return {
    type: SET_ONLINE_STATUS,
    payload: { isOnline },
  };
}
