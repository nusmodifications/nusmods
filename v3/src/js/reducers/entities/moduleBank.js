// @flow
import type { FSA } from 'types/redux';
import type {
  Module,
  ModuleCode,
  ModuleCondensed,
  Semester,
} from 'types/modules';
import type {
  SemTimetableConfig,
} from 'types/timetables';
import type {
  ModuleList,
  ModuleSelectList,
  ModuleSelectListItem,
  ModuleCodeMap,
} from 'types/reducers';

import _ from 'lodash';

import { FETCH_MODULE_LIST, FETCH_MODULE } from 'actions/moduleBank';
import * as RequestResultCases from 'middlewares/requests-middleware';

export type ModulesMap = {
  [ModuleCode]: Module,
};
export type ModuleBank = {
  moduleList: ModuleList,
  modules: ModulesMap,
  moduleSelectList: ModuleSelectList,
  moduleCodes: ModuleCodeMap,
};

const defaultModuleBankState: ModuleBank = {
  moduleList: [], // List of modules
  modules: {}, // Object of ModuleCode -> ModuleDetails
  moduleSelectList: [],
  moduleCodes: {},
};

function precomputeFromModuleList(moduleList: ModuleList) {
  // Cache a mapping of all module codes to module data for fast module data lookup
  const moduleCodes = _.zipObject(
    moduleList.map(module => module.ModuleCode),
    moduleList,
  );

  // Precompute this in reducer because putting this inside render is very expensive (5k modules!)
  const moduleSelectList = moduleList.map((module: ModuleCondensed) => ({
    semesters: module.Semesters,
    value: module.ModuleCode,
    label: `${module.ModuleCode} ${module.ModuleTitle}`,
  }));

  return { moduleCodes, moduleSelectList };
}

function moduleBank(state: ModuleBank = defaultModuleBankState, action: FSA): ModuleBank {
  switch (action.type) {
    case FETCH_MODULE_LIST + RequestResultCases.SUCCESS:
      return {
        ...state,
        ...precomputeFromModuleList(action.payload),
        moduleList: action.payload,
      };

    case FETCH_MODULE + RequestResultCases.SUCCESS:
      return {
        ...state,
        modules: {
          ...state.modules,
          [action.payload.ModuleCode]: action.payload,
        },
      };

    default:
      // FIXME: HACK - Temporary solution to not having a specific dehydration action
      if (!state.moduleCodes && state.moduleList) {
        return {
          ...state,
          ...precomputeFromModuleList(state.moduleList),
        };
      }

      return state;
  }
}

export function getSemModuleSelectList(
  state: ModuleBank,
  semester: Semester,
  semTimetableConfig: SemTimetableConfig,
): ModuleSelectListItem[] {
  return state.moduleSelectList.filter((item: ModuleSelectListItem): boolean => {
    // In specified semester and not within the timetable.
    return _.includes(item.semesters, semester) && !semTimetableConfig[item.value];
  });
}

export default moduleBank;
