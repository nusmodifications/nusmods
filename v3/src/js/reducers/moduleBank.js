// @flow
import type { FSA } from 'types/redux';
import type { Module, ModuleCode, Semester } from 'types/modules';
import type { SemTimetableConfig } from 'types/timetables';
import type { ModuleList, ModuleSelectListItem, ModuleCodeMap } from 'types/reducers';

import _ from 'lodash';

import { FETCH_MODULE_LIST, FETCH_MODULE } from 'actions/moduleBank';
import * as RequestResultCases from 'middlewares/requests-middleware';

export type ModulesMap = {
  [ModuleCode]: Module,
};
export type ModuleBank = {
  moduleList: ModuleList,
  modules: ModulesMap,
  moduleCodes: ModuleCodeMap,
};

const defaultModuleBankState: ModuleBank = {
  moduleList: [], // List of modules
  modules: {}, // Object of ModuleCode -> ModuleDetails
  moduleCodes: {},
};

function precomputeFromModuleList(moduleList: ModuleList) {
  // Cache a mapping of all module codes to module data for fast module data lookup
  const moduleCodes = _.zipObject(moduleList.map(module => module.ModuleCode), moduleList);

  return { moduleCodes };
}

function moduleBank(state: ModuleBank = defaultModuleBankState, action: FSA): ModuleBank {
  switch (action.type) {
    case FETCH_MODULE_LIST + RequestResultCases.SUCCESS:
      return {
        ...state,
        ...precomputeFromModuleList(action.payload),
        moduleList: action.payload,
      };

    case FETCH_MODULE + RequestResultCases.SUCCESS:
      return {
        ...state,
        modules: {
          ...state.modules,
          [action.payload.ModuleCode]: action.payload,
        },
      };

    default:
      // FIXME: HACK - Temporary solution to not having a specific dehydration action
      if (!state.moduleCodes && state.moduleList) {
        return {
          ...state,
          ...precomputeFromModuleList(state.moduleList),
        };
      }

      return state;
  }
}

export function getSemModuleSelectList(
  state: ModuleBank,
  semester: Semester,
  semTimetableConfig: SemTimetableConfig,
): ModuleSelectListItem[] {
  return state.moduleList
    // In specified semester and not within the timetable.
    .filter(item => item.Semesters.includes(semester))
    .map(mod => ({
      ...mod,
      isAdded: mod.ModuleCode in semTimetableConfig,
    }));
}

export default moduleBank;
