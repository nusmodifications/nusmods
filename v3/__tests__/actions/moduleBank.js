// @flow
import type { FSA } from 'types/redux';

import * as actions from 'actions/moduleBank';

test('fetchModuleList should return a request action', () => {
  const resultOfAction: FSA = actions.fetchModuleList();
  expect(resultOfAction).toMatchSnapshot();
});

test('fetchModule should return a request action', () => {
  const resultOfAction: FSA = actions.fetchModule('CS1010S');
  expect(resultOfAction).toMatchSnapshot();
});

// TODO: Find out how to test this using snapshots
test('loadModule should dispatch a request', () => {
  expect(typeof actions.loadModule('test') === 'function').toBe(true);
});
