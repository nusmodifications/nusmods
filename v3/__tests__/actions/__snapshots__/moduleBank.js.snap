// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchModule should return a request action 1`] = `
Object {
  "meta": Object {
    Symbol(API_REQUEST): true,
  },
  "payload": Object {
    "method": "GET",
    "url": "https://nusmods.com/api/2016-2017/modules/CS1010S.json",
  },
  "type": "FETCH_MODULE",
}
`;

exports[`fetchModuleList should return a request action 1`] = `
Object {
  "meta": Object {
    Symbol(API_REQUEST): true,
  },
  "payload": Object {
    "method": "GET",
    "url": "https://nusmods.com/api/2016-2017/moduleList.json",
  },
  "type": "FETCH_MODULE_LIST",
}
`;

exports[`loadModule should dispatch a request if module is not found 1`] = `
Object {
  "meta": Object {
    Symbol(API_REQUEST): true,
  },
  "payload": Object {
    "method": "GET",
    "url": "https://nusmods.com/api/2016-2017/modules/test.json",
  },
  "type": "FETCH_MODULE",
}
`;
